### PR TITLE
keep toplevel dir modes and symlinks

### LINKED ISSUE #80: sync top level dir permissions
We should make sure top level directories also have the right permissions.
```
ubuntu@pashtestold:~/try$ ./try ls -lah /tmp | head -2
total 1.2M
drwxrwxr-x  1 root   root    4.0K Jun 27 19:03 .

ubuntu@pashtestold:~/try$ ls -lah /tmp | head -2
total 1.3M
drwxrwxrwx 84 root   root    112K Jun 27 19:03 .
```

### LINKED ISSUE #139: top level symlink handled incorrectly
symlinks in the root is being handled as directories.

### PR DESCRIPTION
**Pull request by ezrizhu**
- [X] Modify try to keep toplevel dir perms
- [x] sticky bit
- [x] #139
- [x] add tests (need to ignore file size, swapfile, etc..., and timestamp)

In this pull request, we are ensuring the dirs and files in the `/` root path of the temproot is as close as the 'real' one as possible. 

`$SANDBOX_DIR/temproot` is now being created with the same permission as the host, and every other directories are created with the same mode as the real one.
Symlinks are now also created in the unshare, and removed after unshare finishes. 
Tests are created to check the mode, ownership, and symlink of the files in the `/` directory. 

Known issues
In the test, we're ignoring files with the name swap.
And also /proc, our current `mount -t proc proc /proc` invocation are creating the /proc dir with nobody and nogroup ownership. 

This PR currently assumes there are no other files in the root dir besides the swap.img.

Fixes #80 
Fixes #139

### CODE REVIEW

File: test/reuse_problematic_sandbox.sh [hunk 1]
```diff
@@ -30,5 +30,5 @@ try_example_dir=$(mktemp -d)
 #               at the moment, this modification will be caught as illegal by `try`,
 #               but it doesn't seem to both overlayfs at all.
 # TODO: Extend this with more problematic overlayfs modifications.
-touch "$try_example_dir/temproot/bin/foo"
+touch "$try_example_dir/temproot/etc/foo"
```
[the remark this answers is not in the document]
Comment (ezrizhu) [reply test/reuse_problematic_sandbox.sh:33]: bin is a symlink sometimes, so they will no longer persist after try stops

File: try [hunk 2]
```diff
@@ -214,6 +222,18 @@ do
...
 
+## now we deal with symlinks
+for mountpoint in $(cat "$DIRS_AND_MOUNTS")
+do
+    ## We're only interested in symlinks
+    if ! [ -L "$mountpoint" ]
+    then
+        continue
+    fi
+    #ln -s "$sandbox_dir/temproot/$(readlink "$mountpoint")" "$sandbox_dir/temproot/$mountpoint"
```
[the remark this answers is not in the document]
Comment (ezrizhu) [reply try]: oops, that was for my own comparison purposes while in development, it has been removed.

File: try [hunk 3]
```diff
@@ -214,6 +222,18 @@ do
     fi
 done
 
+## now we deal with symlinks
+for mountpoint in $(cat "$DIRS_AND_MOUNTS")
```
[the remark this answers is not in the document]
Comment (ezrizhu) [reply try]: ah that make sense, I have implemented that in the latest commit.

File: try [hunk 4]
```diff
@@ -99,9 +100,10 @@ try() {
     while IFS="" read -r mountpoint
     do
         ## Only make the directory if the original is a directory too
-        if [ -d "$mountpoint" ]
+        if [ -d "$mountpoint" ] && [ ! -L "$mountpoint" ]
```
Comment (mgree) [inline try:103]: `!` goes outside `[` form

**Review by mgree — CHANGES_REQUESTED** [review]
Nice! Just a few questions.

File: try [hunk 5]
```diff
@@ -214,6 +222,17 @@ do
     fi
 done
 
+## now we deal with symlinks
```
[the remark this answers is not in the document]
Comment (ezrizhu) [reply try]: Good point! This has been fixed in b9a6f6ffbaaa1ef46ccb00a64a4a9d221879f18a :3

File: try [hunk 6]
```diff
@@ -263,6 +282,15 @@ EOF
     unshare --mount --map-root-user --user --pid --fork $EXTRA_NS "$mount_and_execute"
     TRY_EXIT_STATUS=$?
 
+    # remove symlink
```
[the remark this answers is not in the document]
Comment (ezrizhu) [reply try:285]: Currently no modified things in the top level are being sync'd. Should we address this in a separate PR/issue? Working on that may require more engineering efforts. 

We're also only recreating directories and symlinks (only to directories) in the top level. No files are being recreated.

[the remark this answers is not in the document]
Comment (mgree) [reply try:286]: Weird. We should add it (not nec in this PR).

[the remark this answers is not in the document]
Comment (ezrizhu) [reply try:286]: Tracking in #150

File: try [hunk 7]
```diff
@@ -273,6 +283,17 @@ EOF
...
+    # first set temproot to be writible, rhel derivatives defaults / to r-xr-w-rw
```
Comment (mgree) [inline try:287]: Typo here in perms?

**Review by mgree — APPROVED** [review]
Looks good to me (but maybe fix the typo, no need for re-review).

**Pull request merged**